java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UpdateDependencyHint.java

```diff
@@ -0,0 +1,259 @@
+        for (char c : v.toCharArray()) {
+            if (!(Character.isDigit(c) || c == '.')) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean noTimestamp(String v) {
+        char[] chars = v.toCharArray();
+        for (int i = 0; i < chars.length; i++) {
+            if (!(Character.isDigit(chars[i]))) {
+                return i == 0 || Integer.parseInt(v.substring(0, i)) < 10_000;
+            }
+        }
+        return Integer.parseInt(v) < 10_000;
+    }
+
+    // example: in '3.14' -> out '3.'
+    private String getMajorComponentPrefix(String v) {
+        int dot = v.indexOf('.');
+        if (dot > 0) {
+            String major = v.substring(0, dot+1);
+            if (isNumerical(major)) {
+                return major;
+            }
+        }
+        return "";
+    }
+
+    private static int compare(String version1, String version2) {
+        return new ComparableVersion(version1).compareTo(new ComparableVersion(version2));
+    }
+
+    private ErrorDescription createHintForComponent(POMExtensibilityElement comp, String version) {
+        Document doc = comp.getModel().getBaseDocument();
+        int line = NbEditorUtilities.getLine(doc, comp.findPosition(), false).getLineNumber() + 1;
+        List<Fix> fix =  Collections.singletonList(new UpdateVersionFix(comp, version));
+        return ErrorDescriptionFactory.createErrorDescription(Severity.HINT, HINT_UpdateDependencyHint() + version, fix, doc, line);
+    }
+
+    private static class UpdateVersionFix implements Fix {
+
+        private final POMExtensibilityElement version_comp;
+        private final String version;
+
+        private UpdateVersionFix(POMExtensibilityElement component, String toVersion) {
+            this.version_comp = component;
+            this.version = toVersion;
+        }
+
+        @Override
+        public String getText() {
+            return FIX_UpdateDependencyHint() + version;
+        }
+
+        @Override
+        public ChangeInfo implement() throws Exception {
+            PomModelUtils.implementInTransaction(version_comp.getModel(), () -> {
+                version_comp.setElementText(version);
+            });
+            return new ChangeInfo();
+        }
+
+    }
+
+    @Override
+    public void cancel() {
+        customizer = null;
+    }
+
+    @Override
+    public Configuration getConfiguration() {
+        return config;
+    }
+
+    private static boolean getNoMajorUpgradeOption() {
+        return config.getPreferences().getBoolean(KEY_NO_MAJOR_UPGRADE, false);
+    }
+
+    @Override
+    public String getSavedValue(JComponent customizer, String key) {
+        if (KEY_NO_MAJOR_UPGRADE.equals(key) && customizer instanceof UpdateDependencyHintCustomizer) {
+            return Boolean.toString(((UpdateDependencyHintCustomizer)customizer).getSavedNoMajorUpgradeOption());
+        }
+        return null;
+    }
```

Comment:
The maven hints option panel seems to be fairly broken atm. Cancel doesn't work for most hint options etc. The implementation is also a bit... unusual. It polls the `getSavedValue` method twice per second and it forgets polling again when cancelled and reopened.

It also expects the UI to remember the old ("saved") value for the options? I don't think it is right now possible to make the ok, cancel and apply button work at the same time. I chose ok.

Reply:
i believe @neilcsmith-net mentioned this once, now I know what you meant :D

Reply:
I think I have fixed the polling issue. This appears to be the root cause of options not cancelling properly and also the apply button not getting activated. Last commit contains the fix.

Reply:
filed #5165 for the remaining issues